PN532MifareClassic.device.nut

```diff
@@ -0,0 +1,182 @@
+        local frame = _makeMifareFrame(cmd, address, payload);
+        _pn532.sendRequest(frame, responseCallback.bindenv(this), false);
+    }
+    
+    function read(address, callback) {
+        local responseCallback = _getReadCallback(callback);
+        local frame = _makeMifareFrame(CMD_READ_16, address);
+        _pn532.sendRequest(frame, responseCallback.bindenv(this), false);        
+    }
+    
+    function write(address, data, callback) {
+        // Each block has 16 bytes to write to
+        if(data.len() != BLOCK_LENGTH) {
+            imp.wakeup(0, function() {
+                callback(ERROR_INCORRECT_LENGTH, null);
+            });
+        }
+        
+        // Package and send request
+        local responseCallback = _getStatusOkCallback(callback);
+        local frame = _makeMifareFrame(CMD_WRITE_16, address, data);
+        _pn532.sendRequest(frame, responseCallback.bindenv(this), false);      
+    }
+    
+    // -------------------- PRIVATE METHODS -------------------- //
+    
+    static function _makeMifareFrame(command, address, data=null) {
+        local dataLength = data == null ? 0 : data.len();
+        local frame = blob(2 + dataLength);
+        
+        frame.writen(command, 'b');
+        frame.writen(address, 'b');
+        
+        // Optionally add extra data (e.g. write payload)
+        if(data != null) {
+            data.seek(0, 'b');
+            frame.writeblob(data);
+        }
+        
+        // Use the base PN532 class to make the proper frame
+        // We only support communicating with one tag at a time, so target tag 1
+        return PN532.makeDataExchangeFrame(1, frame);
+    }
+    
+    static function _getAuthenticationCallback(userCallback) {
+        return function(error, responseData) {
+            
+            if(error != null) {
+                imp.wakeup(0, function() {
+                    userCallback(error, null);
+                });
+                return;
+            }
+            
+            // Consume response type byte
+            responseData.readn('b');
+            
+            local status = responseData.readn('b');
+            if(status == PN532.STATUS_OK) {
+                imp.wakeup(0, function() {
+                    userCallback(null, true);
+                });
+            } else {
+                local message = status == AUTH_ERROR ? ERROR_BAD_AUTHENTICATION : "Error: " + status;
+                imp.wakeup(0, function() {
+                    userCallback(message, null);
+                });
+            }
+        };
+    }
+    
+    static function _getReadCallback(userCallback) {
+        return function(error, responseData) {
+            
+            if(error != null) {
+                imp.wakeup(0, function() {
+                    userCallback(error, null);
+                });
+                return;
+            }
+            
+            // Consume response type byte
```

Comment:
That sort of interference should be handled at other levels of the comms stack - the PN532 handles RF interference (e.g. the case where multiple tags are in the field) and there is basic checksumming of the packet we receive.

If we're considering the case where there is enough interference to overcome these, should this be something that the library handles in the background before returning the packet to the user?